_rules/element-marked-decorative-has-empty-name-46ca7f.md

```diff
@@ -0,0 +1,157 @@
+---
+id: 46ca7f
+name: Element marked as decorative has empty accessible name
+rule_type: atomic
+description: |
+  This rule checks that elements marked as decorative have an empty accessible name
+accessibility_requirements: # Remove whatever is not applicable
+  wcag20:1.1.1: # Non-text Content (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM tree
+acknowledgments:
+  authors:
+    - Jean-Yves Moyen
+htmlHintIgnore:
+  # https://www.npmjs.com/package/htmlhint
+  # (used with `npm test` to ensure validity of code snippets)
+  - 'alt-require'
+---
+
+## Applicability
+
+The rule applies to any element for which all of the following are true:
+
+- the element is [embedded content][] in the HTML or SVG namespace; and
+- the element is not an `iframe` element; and
+- the element is [marked as decorative][].
+
+**Note:** The list of elements matching the first two conditions is `audio`, `canvas`, `embed`, `img`, `object`, `picture`, `svg`, and `video`. `iframe` elements and `math` elements from the MathML namespace are ignored because they often embed text content.
+
+## Expectation
+
+Each target element as an empty [accessible name][].
+
+## Assumptions
+
+- This rule assumes that the applicable elements embed [non-text content][]. If this is not the case, the rule may fails while [Success Criterion 1.1.1][sc111] is still satisfied because it is only concerned with [non-text content][]. Note that the mismatch between [marking as decorative][marked as decorative] and providing an [accessible name][] is not necessarily an accessibility issue for text content, but is nonetheless bad practice and should be avoided.
+- This rule assumes that elements which are [marked as decorative][] are [pure decoration][]. If this is not the case, this rule may fail while [Success Criterion 1.1.1][sc111] is still satisfied. Note that [marking as decorative][marked as decorative] an element which is not [pure decoration][] is not necessarily an accessibility issue but is nonetheless bad practice and should be avoided.
```

Comment:
I'm uncomfortable with this assumption. The fact that you can add `aria-label` to an element to override `role=none` is useful for things like accessibility overlays. You could also pretty easily get into a situation like this with components. Here's an example:

```jsx
const MyImg = props => (props.alt
  ? <img src={props.src} alt={props.alt} />
  : <img src={props.src} role="none" />
)
const HelloWorldImg = () => <MyImg aria-label="hello world" src="..." />
```

It would be a completely different story if some AT wound announce the label, and others would not. That would be a rule purely based on limited accessibility support for presentation conflict resolution. We have other rules like that. That seems fine. It seems like presentation conflict resolution is fairly well supported by AT, without that, I don't think we can conclude this actually is an issue.

Reply:
I do not understand what you mean.

Do you mean that the assumption is too broad and that many many pages will fail it, thus making the rule mostly useless?

---
1.1.1 states (last bullet):
> If non-text content is pure decoration, is used only for visual formatting, or is not presented to users, then it is implemented in a way that it can be ignored by assistive technology.

So it is cristal clear that if something is pure decoration but cannot be ignored, it is an issue.

Reply:
Yes, while the rule itself is sound, I think this assumption is not going to hold up in practice .

Reply:
> ```js-jsx
> const MyImg = props => (props.alt
>   ? <img src={props.src} alt={props.alt} />
>   : <img src={props.src} role="none" />
> )
> const HelloWorldImg = () => <MyImg aria-label="hello world" src="..." />
> ```

This component would never receive the `aria-label` prop though as it isn't passed along. 

I'm personally quite comfortable with this assumption, but that's based on my personal opinion on the matter and not data.

Reply:
Sorry, yer right, bug. Let me fix that. You may want to check with other implements if they have rules like this. Chair hat off; I don't think we'll be implementing such a rule in axe-core, although we'll likely create one that outright fails anything in a conflict resolution as a best practice. It is a terrible thing to have in yer code base. I'm confused why ARIA even allows it. I'm going to file an issue about that.

Reply:
> I don't think we'll be implementing such a rule in axe-core, although we'll likely create one that outright fails anything in a conflict resolution as a best practice.

In its current state, failing things triggering conflict resolution should count as an implementation of the rule, as far as I understand (that doesn't fail any of the Passed/Inapplicable test cases, and that doesn't pass any of the Failed test cases).

Failing every conflict resolution will also make you fail, say, `<span role="none" aria-label="foo">foo</span>` which is inapplicable for the rule because it is not non-text content (so 1.1.1 doesn't care).
We can try and make sure to avoid writing examples like that in order to let tools failing any conflict resolution be counted as implementation for the rule.

I do agree with you that 
>  [conflict resolution] is a terrible thing to have in yer code base. I'm confused why ARIA even allows it. 

so I fully support wider implementations of this rule. The restriction to these specific cases is to make the rule map to 1.1.1 and I am not sure we have a similar SC to map conflict resolution of text content (1.3.1 seems a bit far fetched).

The first assumption explicitly states that conflict resolution on text-content is bad practice but not a WCAG issue, the second one states that conflict resolution is bad practice if done on purpose on non-decorative elements. Both are needed to ensure the mapping to 1.1.1 which is likely needed to make this rule publishable. But both are also "if this assumption is false and the rule fails, the page might pass 1.1.1 but is still crap and need to be fixed"…

That's the main reason why I'm happy with these assumptions: when false, the rule won't flag stuff that are perfectly OK. It will flag stuff that are bad practice but ignored by WCAG. So whether these assumptions hold or not, when the rule fail, the code needs to be fixed.